serde-reflection/src/de.rs

```diff
@@ -391,55 +396,147 @@ impl<'de, 'a> de::Deserializer<'de> for Deserializer<'de, 'a> {
+                &mut VariantFormat::unknown(),
+            ));
+        }
+
+        // First visit each of the variants by name according to `variants`.
+        // Later revisit them by u32 index until an index matching each of the
+        // named variants has been determined.
+        let provisional_min = u32::MAX - (variants.len() - 1) as u32;
+        for (i, &variant_name) in variants.iter().enumerate() {
+            if !self
+                .tracer
+                .discriminants
+                .contains_key(&(enum_type_id, VariantId::Name(variant_name)))
+            {
+                // Insert into known_variants with a provisional index.
+                let provisional_index = provisional_min + i as u32;
+                let variant = known_variants
+                    .entry(provisional_index)
+                    .or_insert_with(|| Named {
+                        name: variant_name.to_owned(),
+                        value: VariantFormat::unknown(),
+                    });
+                self.tracer
+                    .incomplete_enums
+                    .insert(enum_name.into(), EnumProgress::NamedVariantsRemaining);
+                // Compute the discriminant and format for this variant.
+                let mut value = variant.value.clone();
+                let enum_value = visitor.visit_enum(EnumDeserializer::new(
+                    self.tracer,
+                    self.samples,
+                    VariantId::Name(variant_name),
+                    &mut value,
+                ))?;
+                let discriminant = Discriminant::of(&enum_value);
+                self.tracer
+                    .discriminants
+                    .insert((enum_type_id, VariantId::Name(variant_name)), discriminant);
+                return Ok(enum_value);
+            }
+        }
+
+        // We know the discriminant for every variant name. Now visit them again
+        // by index to find the u32 id that goes with each name.
+        //
+        // If there are no provisional entries waiting for an index, just go
+        // with index 0.
+        let mut index = 0;
+        if known_variants.range(provisional_min..).next().is_some() {
+            self.tracer
+                .incomplete_enums
+                .insert(enum_name.into(), EnumProgress::IndexedVariantsRemaining);
+            while known_variants.contains_key(&index)
+                && self
+                    .tracer
+                    .discriminants
+                    .contains_key(&(enum_type_id, VariantId::Index(index)))
+            {
+                index += 1;
             }
-            index
+        }
+
+        // Compute the discriminant and format for this variant.
+        let mut value = VariantFormat::unknown();
+        let enum_value = visitor.visit_enum(EnumDeserializer::new(
+            self.tracer,
+            self.samples,
+            VariantId::Index(index),
+            &mut value,
+        ))?;
```

Comment:
That's right. Serde-reflection will try indices `0..` until it has matched each named variant to an index. So if there are `variants.len()` named variants, then there are at least 1 and at most `variants.len()` distinct possible discriminants. If each index deserializes to a unique discriminant (which is always the case in derived impls) then this would never exceed `variants.len() - 1` by trying indices in the `0..` sequence.

There are other weird things that a handwritten `visit_enum` implementation could do: discontiguous index ranges, or multiple indices deserializing to the same variant. Neither of these is supported by the previous implementation either.